include/dice/sparse_map/internal/sparse_hash.hpp

```diff
@@ -0,0 +1,943 @@
+			using type = void;
+			using const_reference = void;
+			using reference = void;
+		};
+
+		template<typename VSel> requires requires { typename VSel::value_type; }
+		struct get_mapped_type<VSel> {
+			using type = typename VSel::value_type;
+			using const_reference = type const &;
+			using reference = type &;
+		};
+
+	public:
+		template<bool IsConst>
+		struct sparse_iterator;
+
+		using key_type = typename KeyValueSelect::key_type;
+		using mapped_type = typename get_mapped_type<KeyValueSelect>::type;
+		using mapped_const_reference = typename get_mapped_type<KeyValueSelect>::const_reference;
+		using mapped_reference = typename get_mapped_type<KeyValueSelect>::reference;
+		using value_type = ValueType;
+		using hasher = Hash;
+		using key_equal = KeyEqual;
+		using allocator_type = Allocator;
+		using growth_policy = GrowthPolicy;
+		using reference = value_type &;
+		using const_reference = value_type const &;
+		using size_type = typename std::allocator_traits<allocator_type>::size_type;
+		using pointer = typename std::allocator_traits<allocator_type>::pointer;
+		using const_pointer = typename std::allocator_traits<allocator_type>::const_pointer;
+		using difference_type = typename std::allocator_traits<allocator_type>::difference_type;
+		using iterator = sparse_iterator<false>;
+		using const_iterator = sparse_iterator<true>;
+
+		static constexpr size_type default_init_bucket_count = 0;
+
+		static constexpr float max_load_factor = static_cast<float>(MaxLoadFactor::num) / static_cast<float>(MaxLoadFactor::den);
+		static_assert(max_load_factor >= 0.1f && max_load_factor <= 0.8f,
+					  "Specified invalid MaxLoadFactor, must be in range [0.1, 0.8]");
+
+	private:
+		[[nodiscard]] static constexpr size_type calc_load_threshold_rehash(size_type bucket_count) noexcept {
+			return size_type(float(bucket_count) * max_load_factor);
+		}
+
+		[[nodiscard]] static constexpr size_type calc_load_threshold_clear_deleted(size_type bucket_count) noexcept {
+			float const max_load_factor_with_deleted_buckets = max_load_factor + 0.5f * (1.0f - max_load_factor);
+			assert(max_load_factor_with_deleted_buckets > 0.0f && max_load_factor_with_deleted_buckets <= 1.0f);
+
+			return size_type(float(bucket_count) * max_load_factor_with_deleted_buckets);
+		}
+
+
+		static constexpr bool has_mapped_type = !std::is_same_v<mapped_type, void>;
+
+		using sparse_bucket_array_type = sparse_bucket_array<value_type, allocator_type, Sparsity>;
+		using sparse_bucket_type = typename sparse_bucket_array_type::bucket_type;
+
+	private:
+		[[no_unique_address]] hasher h_;
+		[[no_unique_address]] key_equal keq_;
+		[[no_unique_address]] growth_policy gpol_;
+
+		sparse_bucket_array_type buckets_;
+		size_type bucket_count_;
+
+		size_type n_elements_;
+		size_type n_deleted_elements_;
+
+		/**
+		 * Maximum that size_ can reach before a rehash occurs automatically
+		 * to grow the hash table.
+		 */
+		size_type load_threshold_rehash_;
```

Comment:
Does this (and the value below) really make a difference? Is it measurable slower if we calculate it on the fly?